src/components/shared/ContractFunction.vue

```diff
@@ -0,0 +1,366 @@
+<script setup lang="ts">
+import useWeb3 from '@/compositions/useWeb3'
+import { BN } from 'bn.js'
+import { reactive, computed, watch, onMounted } from 'vue'
+import { toWei, Unit } from 'web3-utils'
+import { MethodSelect, MethodType } from '../endpoints/SendTransaction.vue'
+import ParamField from './ParamField.vue'
+
+interface ElementType extends MethodType {
+  error?: boolean
+}
+type Props = {
+  call?: string
+  modelValue?: MethodType[]
+  custom?: boolean
+  onlyParams?: boolean
+  hideData?: boolean
+  data?: string
+  testidPrefix?: string
+}
+
+type Emits = {
+  (event: 'update:modelValue', value: any): void
+  (event: 'update:error', value: boolean): void
+  (event: 'update:data', value: string): void
+}
+
+const props = defineProps<Props>()
+const emits = defineEmits<Emits>()
+
+const { getWeb3 } = useWeb3()
+const { eth } = getWeb3()
+
+const SIGNATURE_CACHE = 'signature-cache'
+
+type BytesSignatureResponse = {
+  count: number
+  next: unknown
+  previous: unknown
+  results: [
+    {
+      id: number
+      created_at: string
+      text_signature: string
+      hex_signature: string
+      bytes_signature: string
+    }
+  ]
+}
+
+const fetcher = async <Response, Request>(config: {
+  url: string
+  method: 'GET' | 'POST'
+  data?: Request
+  headers?: Record<string, never>
+}): Promise<Response> => {
+  const fetchConfig: RequestInit = {
+    method: config.method,
+    headers: config.headers || {
+      Accept: 'application/json',
+      'Content-Type': 'application/json',
+    },
+  }
+  if (config.data) {
+    fetchConfig.body = JSON.stringify(config.data)
+  }
+
+  const response = await fetch(config.url, fetchConfig)
+
+  if (!response.ok) {
+    return response
+      .json()
+      .catch(() => {
+        throw new Error(response.status.toString())
+      })
+      .then(message => {
+        throw message
+      })
+  }
+  return await response.json()
+}
+
+const decodeData = async (data: string): Promise<MethodSelect> => {
```

Comment:
Into util

Reply:
I moved both this and the fetcher into a functionUtils helper